CanonicalABI.md

```diff
@@ -0,0 +1,1137 @@
+interleaved.
+
+```
+def lift(direction, src, ty, values):
+  match ty
+    f32 => yield(values.assert_next(wasm_f32)),
+    f64 => yield(values.assert_next(wasm_f64)),
+
+    s8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -128 && val <= 127)
+      yield(val as s8)
+    }
+
+    s16 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -32768 && val <= 32767)
+      yield(val as s16)
+    }
+
+    u8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= 0 && val <= 255)
+      yield(val as u8)
+    }
+    u16 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= 0 && val <= 65535)
+      yield(val as u16)
+    }
+
+    s32 => yield(values.assert_next(wasm_i32) as s32),
+    u32 => yield(values.assert_next(wasm_i32) as u32),
+    s64 => yield(values.assert_next(wasm_i64) as s64),
+    u64 => yield(values.assert_next(wasm_i64) as u64),
+
+    char => {
+      codept = values.assert_next(wasm_i32)
+      # generate a trap if `codept` is out of bounds
+      assert_or_trap(codept < 0xD800 || (codpt >= 0xE000 && codept <= 0x10FFFF))
+      yield(codept as char)
+    }
+
+    # Note that lists are only deallocated during lowering for exports. For
+    # imports it's assumed the memory is still owned by the caller WebAssembly
+    # module.
+    #
+    # Also note that `read_utf8_string` here is expected to validate that the
+    # input read is indeed valid utf-8. One option is to trap on invalid utf-8,
+    # but another option is to match `TextDecoder.decode` on the web with
+    # replacement error mode which produces replacement characters.
+    #
+    # Currently it's expected taht `read_utf8_string` will not trap and will
+    # instead follow the `TextDecoder.decode` semantics which is what WTF-16
+    # languages like JS/Java/C# will want.
+    string => {
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      val = read_utf8_string(src, ptr, len)
+      if direction == "export":
+        src.canonical_abi_free(ptr, len, 1)
+      yield(val as string)
+    }
+
+    list<$ty> => {
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      assert_or_trap(ptr % align(direction, $ty) == 0)
+      yield(list_start<$ty>(len))
+
+      for i in 0..len:
+        yield(lift_from_memory(direction, src, $ty, ptr + i * size(direction, $ty)))
+      if direction == "export":
+        src.canonical_abi_free(ptr, len * size(direction, $ty), align(direction, $ty))
+      yield(list_done<$ty>)
+    }
+
+    flags<$fields> => {
+      result = {}
+      for chunk in $fields.chunks(32):
+        bits = values.assert_next(i32)
+        for i, field in enumerate(chunk):
+          result[field] = bits & (1 << i) != 0
+      yield(result as flags<$fields>)
+    }
+
+    record<$fields> => {
+      yield(record_start<$fields>)
+      for $field in $fields:
+        lift(direction, src, values)
```

Comment:
```suggestion
        yield from lift(direction, src, values)
```

Reply:
And, iiuc, you need to have all other recursive calls to `lift(...)`/`lift_from_memory(...)` be `yield from` to get the desired transitive yield behavior.